File: 20-greedy/grass.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;
typedef long long ll;
typedef unsigned long long ull;
typedef long double lf;
typedef vector<ll> vi;
typedef vector<vi> mi;
typedef pair<ll, ll> ii;
typedef vector<ii> vii;
typedef vector<vii> mii;
const int INF = ~(1<<31);
template <class T> T smod(T a, T b) { return (a % b + b) % b; }

// 1. Transform problem to 1 dimension using pythagoras
// 2. Sort on left
// 3. keep watered
// 4. Find largest current right while curr left <= watered

pair<lf, lf> best_between(ll& i, lf left, lf right, vector<pair<lf,lf>>& intervals) {
  pair<lf, lf> best{-1, -1};
  for (; i < intervals.size() && intervals[i].first <= left; i++)
  {
    best = (best.second > intervals[i].second) ? best : intervals[i];
  }
  return best;
}

ll solve(vector<pair<lf,lf>>& intervals, ll L) {
  ll idx = 0;
  lf left = 0;
  lf right = L;

  ll count = 0;
  while (left < right) {
    auto best = best_between(idx, left, right, intervals);
    if (best.first == -1 && best.second == -1) return -1;
    count++;
    left = best.second;
  }

  return count;
}

int main(void) {
  ll N, L, W;
  while (cin >> N >> L >> W) {

    vector<pair<lf, lf>> intervals;

    ll x, r;
    for (ll i = 0; i < N; i++)
    {
      cin >> x >> r;
      lf side_length = sqrt((r*r) - ((W/2.0)*(W/2.0)));
      lf left = x-side_length;
      lf right = x+side_length;
      cout << left << " " << right << endl;
      intervals.push_back(pair<lf, lf>{left, right});
    }

    sort(intervals.begin(), intervals.end());

    cout << solve(intervals, L) << endl;
  }
  return 0;
}

File: 20-greedy/grass.py
from decimal import Decimal
import sys
import math

# 1. Transform problem to 1 dimension using pythagoras
# 2. Sort on left
# 3. keep watered
# 4. Find largest current right while curr left <= watered

def best_between(i, left, intervals):
  best = (-1, -1)
  while i < len(intervals) and intervals[i][0] <= left:
    best = best if best[1] > intervals[i][1] else intervals[i]
    i += 1
  return (i, best)

def solve(intervals, L):
  idx = 0
  left = 0
  right = L

  count = 0
  while (left < right):
    (i, best) = best_between(idx, left, intervals);
    idx = i
    if (best[0] == -1 and best[1] == -1):
      return -1
    count += 1
    left = best[1]

  return count

for line in sys.stdin:
  N, L, W = map(int, line.split())
  intervals = []

  for i in range(N):
    x, r = map(int, input().split())
    pW = Decimal(W)
    px = Decimal(x)
    pr = Decimal(r)
    p2 = Decimal(2)
    side_length = Decimal(math.sqrt(Decimal(pr**2) - Decimal((pW/p2)**2)));
    left = px-side_length;
    right = px+side_length;
    intervals.append((left, right));

  intervals.sort()

  print(solve(intervals, L))
